synapse/storage/events.py

```diff
@@ -2037,13 +2038,33 @@ def _purge_history_txn(
                 retcols=["state_group"],
                 keyvalues={},
             )
-            remaining_state_groups.extend(
-                row["state_group"] for row in rows
 
-                # exclude state groups we are about to delete: no point in
-                # updating them
-                if row["state_group"] not in state_groups_to_delete
-            )
+            for row in rows:
+                sg = row["state_group"]
+
+                if sg in state_groups_to_delete:
+                    # exclude state groups we are about to delete: no point in
+                    # updating them
+                    continue
+
+                if not self._is_state_group_referenced(txn, sg):
```

Comment:
if you're going to reference methods in StateGroupWorkerStore, can you make EventsStore inherit from it?

Reply:
Oh, yes. I just assumed we already were (since we use functions from there elsewhere I think).